pkg/cloud/services/compute/instance_test.go

```diff
@@ -728,13 +788,64 @@ func TestService_CreateInstance(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name:                "Delete trunks on port creation error",
+			getMachine:          getDefaultMachine,
+			getOpenStackCluster: getDefaultOpenStackCluster,
+			getOpenStackMachine: func() *infrav1.OpenStackMachine {
+				m := getDefaultOpenStackMachine()
+				m.Spec.Ports = []infrav1.PortOpts{
+					{Description: "Test port 0", Trunk: pointer.BoolPtr(true)},
+					{Description: "Test port 1"},
+				}
+				return m
+			},
+			expect: func(computeRecorder *MockClientMockRecorder, networkRecorder *mock_networking.MockNetworkClientMockRecorder) {
+				extensions := []extensions.Extension{
+					{Extension: common.Extension{Alias: "trunk"}},
+				}
+				networkRecorder.ListExtensions().Return(extensions, nil)
+
+				// BUG! Port must not exist already or we won't create the trunk
+				expectCreatePort(networkRecorder, portName, networkUUID)
+
+				// Check for existing trunk
+				networkRecorder.ListTrunk(newGomegaMockMatcher(
+					MatchFields(IgnoreExtras, Fields{
+						"Name":   Equal(portName),
+						"PortID": Equal(portUUID),
+					}),
+				)).Return([]trunks.Trunk{}, nil)
+
+				// Create new trunk
+				networkRecorder.CreateTrunk(newGomegaMockMatcher(MatchFields(IgnoreExtras, Fields{
+					"Name":   Equal(portName),
+					"PortID": Equal(portUUID),
+				}))).Return(&trunks.Trunk{
+					PortID: portUUID,
+					ID:     trunkUUID,
+				}, nil)
+				networkRecorder.ReplaceAllAttributesTags("trunks", trunkUUID, attributestags.ReplaceAllOpts{Tags: []string{"test-tag"}}).Return(nil, nil)
+
+				// Looking up the second port fails
+				networkRecorder.ListPort(ports.ListOpts{
+					Name:      "test-openstack-machine-1",
+					NetworkID: networkUUID,
+				}).Return(nil, fmt.Errorf("test error"))
+
+				// We should cleanup the first port and its trunk
+				networkRecorder.DeleteTrunk(trunkUUID).Return(nil)
```

Comment:
controller.go:137: missing call(s) to *mock_networking.MockNetworkClient.DeleteTrunk(is equal to 2cf74a9f-3e89-4546-9779-20f2503c4ae8 (string)) /home/prow/go/src/sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/compute/instance_test.go:837

ok, here's the error and it's expected and it's caused by line 834 above

I agree this is reasonable that we should handle ,however, we have various logic after port creation
so , will all those places need to be handles or as you mentioned, put those stuffs into `status` field
so that we can avoid lost info and delete them in a whole bunch? I actually prefer to this way...thoughts?

Reply:
ok, saw 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/pkg/cloud/services/compute/instance.go#L203

and seems if we add some logic there, it should be good enough

Reply:
@mdbooth I assume if we do some error handling here 

e.g add delete trunk might help us 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/pkg/cloud/services/compute/instance.go#L213

see https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1095 for some discussion points..